src/common/marlin_server.cpp

```diff
@@ -1557,9 +1555,10 @@ void fsm_destroy(ClientFSM type) {
 
 void _fsm_change(ClientFSM type, fsm::BaseData data) {
     {
-        static int previous_type = -1;
-        static int previous_phase = -1;
-        if (previous_type != static_cast<int>(type) || previous_phase != static_cast<int>(data.GetPhase())) {
+        static std::optional<int> previous_type;
+        static std::optional<int> previous_phase;
+        // no value: comparison returns true
+        if (previous_type.value() != static_cast<int>(type) || previous_phase.value() != static_cast<int>(data.GetPhase())) {
```

Comment:
Oops, I probably wasn't clear previously :-|.

The `.value()` is extracting the held value of type `T` and if it's not there it throws an exception (which will likely result in termination, since we have exceptions disabled). That's not what we want.

But there's an overloaded `optional<T> != T` operator out there (eg. if you _don't_ unpack it, just `previous_type != static_cast<int>(type)`, without unpacking.